### PR TITLE
Updated all code for Gnome 3.36, fixes and improvements

### PR DESCRIPTION
**Pull request by lfom**
1. Some refactoring / code cleanup
2. All code updated to Gnome 3.36 (removed all deprecated warnings)
3. Fix: extension can be actually disabled now
4. No animation at all when Animations is disabled in Gnome Tweaks
5. "Half" animation when scrolling in overview (I wanted to make a full loop but failed)

### CODE REVIEW

File: extension.js [hunk 1]
```diff
@@ -64,40 +67,32 @@ function checkSettings() {
...
 function disable() {
+    // log('OverviewAllWindows disable');
     while (signals.length) {
         settings.disconnect(signals.pop());
     }
 
     while (keybindings.length) {
         _removeKeybindings(keybindings.pop());
     }
+    destroy();
```
Comment (amiller27) [inline extension.js]: It seems like if we're going to call this here, then we should also be calling `setUp` in `enable`?

File: workspace.js [hunk 2]
```diff
@@ -2,10 +2,10 @@ const GLib = imports.gi.GLib;
 const Clutter = imports.gi.Clutter;
 const Lang = imports.lang;
 const Main = imports.ui.main;
-const Mainloop = imports.mainloop;
 const Overview = imports.ui.overview;
-const Tweener = imports.ui.tweener;
 const Workspace = imports.ui.workspace;
+// const Mainloop = imports.mainloop;
```
Comment (amiller27) [inline workspace.js]: Can we just delete these if they're unused

File: unifiedWorkspacesView.js [hunk 3]
```diff
@@ -2,49 +2,65 @@
 
 //The following is my own version of WorkspacesView.WorkspacesView
 //It is designed to work with only one "Workspace" containing all windows
-
-const { Clutter, Gio, GObject, Meta, Shell, St } = imports.gi;
+const { Clutter, GObject, Meta, Shell, St } = imports.gi;
 const Signals = imports.signals;
 
+// const Tweener = imports.ui.tweener;
```
Comment (amiller27) [inline unifiedWorkspacesView.js]: Just delete please

File: unifiedWorkspacesView.js [hunk 4]
```diff
@@ -2,49 +2,65 @@
...
 
 var UnifiedWorkspacesView = GObject.registerClass(
 class UnifiedWorkspacesView extends WorkspacesView.WorkspacesViewBase {
     _init(monitorIndex, scrollAdjustment) {
+        super._init(monitorIndex);
+        
         let workspaceManager = global.workspace_manager;
 
-        super._init(monitorIndex);
+        for (let functionName of UnifiedWorkspace.replacedFunctions) {
```
Comment (amiller27) [inline unifiedWorkspacesView.js]: What's the reasoning for moving this in here instead of in extension.js?

**Review by amiller27 — COMMENTED** [review]
Thanks for the PR!  Sorry for not reviewing sooner, I've been busy.  Two bits of feedback other than the comments I left on the code:
About (3), you're right that it seems like disabling is broken on master, but testing your branch it seems like disabling is also broken, at least in some cases?  Specifically when the "Replace Overview" setting is on, if I restart gnome-shell with the extension turned on, and then I disable the extension, it still shows all windows in the overview until I restart gnome-shell.  I'm probably fine to merge without this fixed since it seems like it's broken on master and yours seems less broken, just through I'd ask.
And about (5), I thought the existing version on master had the full loop?  At least it seems like it on my machine.

I also noticed the zoom to/from the overview looks a bit different on your branch - on master the intention is for windows on non-active workspaces to fade in/out instead of zooming to/from their positions on their workspace, if they zoom from their positions like it seems they are on your branch they appear and disappear instantly at the non-overview end of the animation which is a bit jarring

File: extension.js (hunk 1 above)
Comment (lfom) [reply extension.js]: ~~Ooops... Yes, I missed that. Weird it seemed to work fine tho, but this probably explains other things that didn't work as I expected.~~
I didn't fully understand the logic when I first changed the code. Now I know what you mean: the original code always destroys the custom overview, so destroying it again here is not needed (EDIT: got it now, only when using the keybinding it is not needed).

File: workspace.js (hunk 2 above)
Comment (lfom) [reply workspace.js]: Sure, np.

File: unifiedWorkspacesView.js (hunk 3 above)
Comment (lfom) [reply unifiedWorkspacesView.js]: Np

File: unifiedWorkspacesView.js (hunk 4 above)
Comment (lfom) [reply unifiedWorkspacesView.js]: I think I just wanted to clean up the code a bit, and as it seemed to be related to this class I moved the code. Now I wonder if `_init` is being actually called.

**Comment by lfom** [comment]
(3) I think that it is probably related to the fact that I missed calling `setUp` properly. But I kept gnome-shell log running and I didn't see any errors, but I think you are right. I was learning how to code in JS and GS extensions, so probably I missed some details. I am running GS 3.36.4 by the way.

(5) Yeah, probably it was working like this already since I didn't change or added code to override how the windows are animated as I didn't know how to do it then.

I had problems using the extension when I first installed it because I was testing/coding a lot of GS extensions then, and the fact that it would crash when disabled made it impossible to use. Then I guess I didn't pay much attention to how it worked at first and when I tried to do a full loop when changing workspace I disabled part of the code. I can try to revert this, but I am not sure if I can dot it soon.

I will fix the other stuff and do more tests, then I should add another commit.

**Comment by lfom** [comment]
I have downloaded and installed the current version from master trying to figure out how the animation/fade/zoom works, and it does not crash... Then I guess I was using an older version from Extensions site, maybe? It is really weird, since it does not crash and I don't see any errors in the log.

But now I remember why I changed the animation code: when selecting a window from a workspace that isn't the active one, the animation is weird (looks like the window is zoomed and also scrolled at the same time, something like this). And as the animation is also executed even when animations is disabled in Gnome Tweaks, probably that's why I changed the code. Now I don't know what is better: keep the new code related to the animation or not...

**Comment by lfom** [comment]
Well, I think I got it now.... With the current master branch version, this is shown in the log (again, I am using Gnome Shell 3.36.4):

`JS WARNING: [/home/lfom/.local/share/gnome-shell/extensions/OverviewAllWindows@amiller27/extension.js 100]: "prototype" is read-only
`

Then I guess the code from `UnifiedWorkspacesView` is not executed in this case. This explains why I do not see errors in the log. and why you still see the fade effect. Using the same version there is a problem when moving windows while in Overview too: the view ins't updated correctly (the moved window is gone, even tho it still should show all windows).

**Comment by amiller27** [comment]
So I think I follow what you're saying.  For the animation stuff, you're right that essentially 2 things are broken on master:
- Stuff animating when animations are turned off
- Animating when clicking on a window from another workspace in the overview
I think it seems like the first one is fixed on your version; for the second one, I think what's happening is it's zooming from its location on the other workspace or something instead of its location in the overview.

Also you're definitely right that there's the prototype warning on master, if you have a way to fix that that works with everything else that would be nice.  JS really isn't my area either, you very well could know more JS than I do.

Basically feel free to come up with a version that fixes any of the issues you've discovered even if you can't fix all of them, as long as it doesn't break anything already working on master

**Comment by lfom** [comment]
The last commit fixed all problems you have reported previously. Maybe merge the PR to fix the problem disabling the extension (it is a memory leak anyway), and improve the animations later? The new code does not cause errors, and it is kinda easy to see what changed, so maybe you can re-implement the animations like you planed them to be initially... Unfortunately I haven't used the extension before Gnome 3.36 so I do not know exactly what it was supposed to be.

By the other hand, now that I understood better what is actually going on in the code and I have a better idea on where/how I can look for more info, I will try to disable the unnecessary code I have added so the scroll between workspaces work better (your current master branch code actually does a full loop of the windows, what I think that looks much nicer and it is the expected behavior since this is what happens with the extension disabled). I will try to see if I can improve what happens when activating a window from a workspace that is not the active one too. I can send a new PR later if I manage to fix what is needed.

**Comment by amiller27** [comment]
I don't think I can merge it as-is.  The average user (including myself) on a typical day isn't disabling and re-enabling the extension all the time and won't see the warnings in the logs, but they do see the animations.  If you can separate out any changes to animation from the other fixes so that the animation works as it does on master on gnome-shell 3.36, that would be great, and I'd be happy to merge - I tried just reverting most of the changes to `unifiedWorkspacesView.js` and that didn't do it.

Also one other thing to note - testing today it seems like the current master version correctly turns off animations?  I know I said the other day that it seemed that was broken on master but I can't reproduce that today - can you confirm that some animations still show when the animations switch is turned off for you on master, and if so, any steps required to reproduce?

**Comment by lfom** [comment]
Even if you do not disable the extension manually, it is disabled every time the screen is locked or the computer suspended.

New version improve the animations a lot, it is much better now: in an empty workspace all windows fade in/out, no bugs moving windows manually, activating windows from a workspace that isn't the active one works better.

**Comment by amiller27** [comment]
Definitely better, thanks.  Still noticed that it isn't quite a full loop when switching between workspaces in the overview - looks like the windows blink out of existence and then scroll in from the edge of the screen - can't really merge until that's fixed (the full loop works right on master).
